app/models/manageiq/providers/cloud_manager/provision/state_machine.rb

```diff
@@ -6,11 +6,37 @@ def create_destination
   def determine_placement
     availability_zone = placement
     options[:dest_availability_zone] = [availability_zone.try(:id), availability_zone.try(:name)]
-    signal :prepare_provision
+    signal :prepare_volumes
+  end
+
+  def prepare_volumes
+    # volumes example: [{name: "vol_name", size: 1}]
+    if miq_request.options[:volumes]
+      phase_context[:requested_volumes] = create_requested_volumes(miq_request.options[:volumes])
+      signal :poll_volumes_complete
+    else
+      signal :prepare_provision
+    end
+  end
+
+  def poll_volumes_complete
+    status, status_message = do_volume_creation_check(phase_context[:requested_volumes])
+    status_message = "completed prepare provision work queued" if status
+    message = "Volume creation is #{status_message}"
+    _log.info(message)
+    update_and_notify_parent(:message => message)
+    if status
+      signal :prepare_provision
+    else
+      requeue_phase
+    end
   end
 
   def start_clone_task
     update_and_notify_parent(:message => "Starting Clone of #{clone_direction}")
+    if phase_context[:requested_volumes]
+      phase_context[:clone_options][:block_device_mapping_v2] = phase_context[:requested_volumes]
```

Comment:
@mansam @Ladas @aufi We start from the classic Provisioning Instance from where we select an image, in properties tab, we have a new section for adding a new disk from a volume. It can be:
- Attaching an existing volume
- Creating and attaching the volume (does this scenario is feasible)

I think this Add Disk during provisioning is currently implemented for Vmware also, we may want to look at this UI part. @dclarizio could you confirm?